File: ios/App/Snappet/Services/KilterLocationService.swift
import CoreLocation
import Foundation

/// A thin, when-in-use CoreLocation wrapper that publishes the device's **current coarse place** for the
/// pre-connect arrival suggestion (Flow 1, Option B). It is deliberately minimal: a single coarse fix
/// (reduced accuracy), bucketed immediately into a `CoarsePlace`, and **never reverse-geocoded, never
/// networked** — the coarse value stays on the device. Authorization is requested **lazily** (only when
/// the feature is actually used), and every denied / restricted / unavailable path degrades to a `nil`
/// place so the module falls back to BLE-only with no crash and no prompt loop.
///
/// All CoreLocation I/O lives here behind the Services edge; the pure place-match (`KilterPlaceMatcher`,
/// in `KilterBoardMemory.swift`) is what the unit tests exercise. The live fix is device-pending.
@MainActor
@Observable
final class KilterLocationService: NSObject {
    /// Authorization state, mapped to a small enum the view can switch on without importing CoreLocation.
    enum Authorization: Equatable {
        case notDetermined   // never asked — the lazy `requestIfNeeded` will prompt
        case denied          // user said no, or it's restricted/unavailable → BLE-only, no prompts
        case authorized      // when-in-use granted
    }

    /// The most recent coarse place, or `nil` when location is unavailable / not yet resolved. Bucketed
    /// (≈110 m) the instant a fix arrives; the precise `CLLocation` is dropped immediately.
    private(set) var currentPlace: CoarsePlace?
    private(set) var authorization: Authorization = .notDetermined

    /// Whether the feature is usable — derived **live** from the current authorization (the view hides
    /// the feature entirely when location is denied/restricted, rather than show a dead toggle). We do
    /// NOT freeze this at init: the global `CLLocationManager.locationServicesEnabled()` is documented as
    /// main-thread-blocking, and CoreLocation already reports a services-off device through the
    /// denied/failure delegate paths, so gating on the status is sufficient and never blocks.
    var isSupported: Bool { authorization != .denied }

    private let manager: CLLocationManager
    /// Guards the single one-shot retry in `didFailWithError` so a failing fix can't trigger a storm.
    /// Main-actor isolated (every read/write is inside a main-actor method or `MainActor.assumeIsolated`);
    /// not UI state, so it's excluded from observation.
    @ObservationIgnored private var didRetryThisRequest = false

    override init() {
        manager = CLLocationManager()
        super.init()
        manager.delegate = self
        // Coarse by design: never ask for precise location for a gym-bucket match.
        manager.desiredAccuracy = kCLLocationAccuracyReduced
        authorization = Self.map(manager.authorizationStatus)
    }

    /// Lazily request when-in-use authorization the first time the feature wants a fix (so the very first
    /// app launch never prompts). When already authorized, also asks for one fresh coarse fix. A no-op
    /// once denied/restricted (no prompt loop) — degrades silently to BLE-only.
    func requestIfNeeded() {
        switch manager.authorizationStatus {
        case .notDetermined:
            manager.requestWhenInUseAuthorization()
        case .authorizedWhenInUse, .authorizedAlways:
            requestFix()
        default:
            break   // denied / restricted — degrade silently to BLE-only
        }
    }

    /// Ask for one fresh coarse fix (e.g. on entering Kilter). No-op unless authorized.
    func refresh() {
        guard manager.authorizationStatus == .authorizedWhenInUse
                  || manager.authorizationStatus == .authorizedAlways else { return }
        requestFix()
    }

    /// Start a one-shot fix request, resetting the per-request retry guard so a fresh request is allowed
    /// its single delayed retry again.
    private func requestFix() {
        didRetryThisRequest = false
        manager.requestLocation()
    }

    private static func map(_ status: CLAuthorizationStatus) -> Authorization {
        switch status {
        case .authorizedWhenInUse, .authorizedAlways: return .authorized
        case .notDetermined: return .notDetermined
        default: return .denied   // .denied / .restricted
        }
    }
}

extension KilterLocationService: CLLocationManagerDelegate {
    nonisolated func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        nonisolated(unsafe) let manager = manager
        MainActor.assumeIsolated {
            authorization = Self.map(manager.authorizationStatus)
            // Newly granted → grab the first coarse fix so the suggestion can appear without another tap.
            if authorization == .authorized { manager.requestLocation() }
            if authorization == .denied { currentPlace = nil }
        }
    }

    nonisolated func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        // Bucket immediately and drop the precise coordinate — only the coarse square is ever kept.
        guard let coord = locations.last?.coordinate else { return }
        let bucket = CoarsePlace(latitude: coord.latitude, longitude: coord.longitude)
        MainActor.assumeIsolated {
            currentPlace = bucket
            // A fix landed — clear the retry guard so the next request gets its retry budget back.
            didRetryThisRequest = false
        }
    }

    nonisolated func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        // A failed fix is not fatal — keep the last known place (if any) and stay BLE-capable. A first
        // failure is often a transient indoor miss, so allow ONE bounded, delayed retry before giving up
        // (no retry storm — the per-request guard permits exactly one).
        nonisolated(unsafe) let manager = manager
        MainActor.assumeIsolated {
            guard !didRetryThisRequest,
                  manager.authorizationStatus == .authorizedWhenInUse
                      || manager.authorizationStatus == .authorizedAlways else { return }
            didRetryThisRequest = true
            DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak manager] in
                manager?.requestLocation()
            }
        }
    }
}

File: android/app/src/main/java/com/snappet/mobile/feature/kilter/KilterLocationService.kt
package com.snappet.mobile.feature.kilter

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.ContextCompat
import com.google.android.gms.location.CurrentLocationRequest
import com.google.android.gms.location.Granularity
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority

/**
 * A thin, when-in-use [com.google.android.gms.location.FusedLocationProviderClient] wrapper that
 * publishes the device's **current coarse place** for the pre-connect arrival suggestion (Flow 1,
 * Option B) — the Android port of the iOS `KilterLocationService`. It is deliberately minimal: a single
 * coarse fix ([Granularity.GRANULARITY_COARSE] / [Priority.PRIORITY_BALANCED_POWER_ACCURACY]), bucketed
 * immediately into a [CoarsePlace], and **never reverse-geocoded, never networked** — the coarse value
 * stays on the device.
 *
 * Authorization is requested **lazily** (only when the feature is actually used, via the caller's
 * permission launcher) and every denied / unavailable path degrades to a `null` place so the module falls
 * back to BLE-only with no crash and no prompt loop. Compose-observable like [KilterBoardController] so the
 * root/Settings recompose when the place or authorization changes.
 *
 * ⚠️ The live FusedLocation fix is device/emulator-pending (like iOS); the pure place bucketing
 * ([CoarsePlace]) + place match ([KilterBoardMemoryRules.suggestion]) are what the unit tests exercise.
 */
class KilterLocationService(context: Context) {

    /** Authorization state, mapped to a small enum the UI can switch on without importing the framework. */
    enum class Authorization { NOT_DETERMINED, DENIED, AUTHORIZED }

    /** The most recent coarse place, or `null` when location is unavailable / not yet resolved. Bucketed
     *  (≈110 m) the instant a fix arrives; the precise coordinate is dropped immediately. */
    var currentPlace by mutableStateOf<CoarsePlace?>(null)
        private set

    var authorization by mutableStateOf(Authorization.NOT_DETERMINED)
        private set

    /** Whether the feature is usable — derived from the current authorization (the UI hides the feature
     *  entirely when denied, rather than show a dead toggle). */
    val isSupported: Boolean get() = authorization != Authorization.DENIED

    private val appContext = context.applicationContext
    private val client = LocationServices.getFusedLocationProviderClient(appContext)
    /** Guards the single one-shot retry per request so a failing fix can't trigger a storm. */
    private var didRetryThisRequest = false

    init {
        authorization = if (hasCoarsePermission()) Authorization.AUTHORIZED else Authorization.NOT_DETERMINED
    }

    /** Whether the COARSE (or finer) location permission is currently granted. */
    private fun hasCoarsePermission(): Boolean =
        ContextCompat.checkSelfPermission(appContext, Manifest.permission.ACCESS_COARSE_LOCATION) ==
            PackageManager.PERMISSION_GRANTED ||
        ContextCompat.checkSelfPermission(appContext, Manifest.permission.ACCESS_FINE_LOCATION) ==
            PackageManager.PERMISSION_GRANTED

    /**
     * Re-read the live permission state and, when authorized, grab one fresh coarse fix. Call this on entry
     * (after the optional runtime request resolves) so the suggestion can appear without another tap. The
     * runtime *prompt* itself is the caller's responsibility (Compose's permission launcher); this never
     * prompts, so it's a safe no-op once denied — no prompt loop.
     */
    fun refresh() {
        authorization = if (hasCoarsePermission()) Authorization.AUTHORIZED else
            // Distinguish "never asked" from "said no": only the caller's launcher result flips DENIED, so
            // before any ask we stay NOT_DETERMINED; after a denied ask the caller calls [onPermissionResult].
            if (authorization == Authorization.DENIED) Authorization.DENIED else Authorization.NOT_DETERMINED
        if (authorization == Authorization.AUTHORIZED) requestFix()
    }

    /**
     * Fold a runtime-permission launcher result into the authorization state (the caller drives the actual
     * `ActivityResultContracts.RequestPermission` prompt). Granted → grab the first coarse fix so the
     * suggestion can appear without another tap; denied → BLE-only, place cleared, no further prompts.
     */
    fun onPermissionResult(granted: Boolean) {
        if (granted) {
            authorization = Authorization.AUTHORIZED
            requestFix()
        } else {
            authorization = Authorization.DENIED
            currentPlace = null
        }
    }

    /** Start a one-shot coarse fix, resetting the per-request retry guard. No-op unless authorized. */
    @SuppressLint("MissingPermission")
    private fun requestFix() {
        if (!hasCoarsePermission()) return
        didRetryThisRequest = false
        val request = CurrentLocationRequest.Builder()
            // Coarse by design: never ask for precise location for a gym-bucket match.
            .setGranularity(Granularity.GRANULARITY_COARSE)
            .setPriority(Priority.PRIORITY_BALANCED_POWER_ACCURACY)
            .build()
        client.getCurrentLocation(request, null)
            .addOnSuccessListener { location ->
                if (location == null) { retryOnce(); return@addOnSuccessListener }
                // Bucket immediately and drop the precise coordinate — only the coarse square is ever kept.
                currentPlace = CoarsePlace.bucket(location.latitude, location.longitude)
                didRetryThisRequest = false
            }
            .addOnFailureListener {
                // A failed fix is not fatal — keep the last known place (if any) and stay BLE-capable.
                retryOnce()
            }
    }

    /** Allow exactly one bounded retry per request (a first failure is often a transient indoor miss). */
    @SuppressLint("MissingPermission")
    private fun retryOnce() {
        if (didRetryThisRequest || !hasCoarsePermission()) return
        didRetryThisRequest = true
        val request = CurrentLocationRequest.Builder()
            .setGranularity(Granularity.GRANULARITY_COARSE)
            .setPriority(Priority.PRIORITY_BALANCED_POWER_ACCURACY)
            .build()
        client.getCurrentLocation(request, null).addOnSuccessListener { location ->
            if (location != null) currentPlace = CoarsePlace.bucket(location.latitude, location.longitude)
        }
    }

    companion object {
        /** The runtime permission the caller's launcher requests for the arrival suggestion. */
        const val PERMISSION = Manifest.permission.ACCESS_COARSE_LOCATION
    }
}
